File: tune/src/tonesrc_test.c
/*
 * Copyright (c) 2017, Intel Corporation
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *   * Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *   * Redistributions in binary form must reproduce the above copyright
 *     notice, this list of conditions and the following disclaimer in the
 *     documentation and/or other materials provided with the distribution.
 *   * Neither the name of the Intel Corporation nor the
 *     names of its contributors may be used to endorse or promote products
 *     derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *
 * Author: Seppo Ingalsuo <seppo.ingalsuo@linux.intel.com>
 *         Liam Girdwood <liam.r.girdwood@linux.intel.com>
 *         Keyon Jie <yang.jie@linux.intel.comel.com>
 */

#include "common_test.h"
#include <reef/ipc.h>
//#include <uapi/tone_abi.h>

/* Keep this 1 to apply tone parameters from command line. Otherwise
 * default tone parameters are used.
 */
#define SEND_COMP_CMD_TONE 0

/* Internal buffer size */
#define TEST_BENCH_DEADLINE 1000 /* 1 ms */
#define TEST_BENCH_NCH 2 /* Stereo */
#define TEST_BENCH_RATE_MAX 48000 /* Hz */
#define TEST_BENCH_PERIOD_FRAMES (TEST_BENCH_RATE_MAX*TEST_BENCH_DEADLINE/1000000)
#define TEST_BENCH_FRAME_SIZE (4 * TEST_BENCH_NCH) /* 4bytes x  NCH */
#define INT_PERIOD_SIZE (TEST_BENCH_PERIOD_FRAMES * TEST_BENCH_FRAME_SIZE)

/* main firmware context */
static struct reef reef;

/*
 * Test Pipeline
 *
 * Tone(0) --- B(3) ---> SRC(1) --B(4)---> Filewrite(2)
 */

/*
 * Components used in static pipeline
 */

#define TONE_ID 0
#define SRC_ID 1
#define FW_ID 2
#define PIPE_ID 5
#define SCHED_ID TONE_ID /* Component ID=0 schedules pipeline */

static struct sof_ipc_comp_tone tone_p0[] = {
	/* ID = 0 */
	SPIPE_GENERIC(SPIPE_COMP(TONE_ID, SOF_COMP_TONE, sof_ipc_comp_tone)),
};

static struct sof_ipc_comp_src src_p0[] = {
	SPIPE_GENERIC(SPIPE_COMP(SRC_ID, SOF_COMP_SRC, sof_ipc_comp_src)), /* ID = 1 */
};

static struct sof_ipc_comp_filewrite filewrite_p0[] = {
	SPIPE_FILEWRITE(SPIPE_COMP(FW_ID, SOF_COMP_FILEWRITE,
	sof_ipc_comp_filewrite),
	"tonesrc_out.txt"), /* ID = 2 */
};

static struct scomps pipe0_scomps[] = {
	SCOMP(tone_p0),
	SCOMP(src_p0),
	SCOMP(filewrite_p0),
};

/*
 * Buffers used in static pipeline
 */
static struct sof_ipc_buffer buffer0[] = {
	/* B3  Tone0 -> SRC1 */
	SPIPE_BUFFER(3, INT_PERIOD_SIZE * 5),

	/* B4  SRC1 -> Filewrite1 */
	SPIPE_BUFFER(4, INT_PERIOD_SIZE * 5),
};

/* pipeline 0 component/buffer connections */
static struct sof_ipc_pipe_comp_connect c_connect0[] = {
	SPIPE_COMP_CONNECT(0, 3), /* p(0): Tone(0) -> B(3) */
	SPIPE_COMP_CONNECT(3, 1), /* p(0): B(3) ->  SRC(1) */
	SPIPE_COMP_CONNECT(1, 4), /* p(0): SRC(1) -> B(4) */
	SPIPE_COMP_CONNECT(4, 2), /* p(0): B(4) ->  Filewrite(2) */
};

/* the static pipelines */
static struct spipe spipe[] = {
	SPIPE(pipe0_scomps, buffer0, c_connect0),
};

/* pipeline, ID=5 */
struct sof_ipc_pipe_new pipeline[] = {
	SPIPE_PIPE(0, 0, PIPE_ID, SCHED_ID, TEST_BENCH_DEADLINE, TASK_PRI_LOW, \
		TEST_BENCH_PERIOD_FRAMES),
};

void usage(char *executable);

#if 0
static int tone_cmd_helper(struct comp_dev *cd, struct sof_ipc_ctrl_data *cmd, int index, int nch, int32_t value)
{
	int i;
	struct sof_ipc_ctrl_value_comp *cdata = (struct sof_ipc_ctrl_value_comp *) cmd->data->data;


	cmd->index = index;
	for (i = 0; i < nch; i++) {
		cdata[i].index = i;
		cdata[i].svalue = value;
	}

	return comp_cmd(cd, COMP_CMD_SET_DATA, cmd);
}
#endif

int main(int argc, char **argv)
{
	struct ipc_comp_dev *pcm_dev;
	struct pipeline *p;
	struct comp_dev *cd;
	struct filewrite_comp_data *fwcd;
	clock_t tic, toc;
	double c_realtime, t_exec;
	int fs1, fs2, n_out, ret, deadline, nch = TEST_BENCH_NCH; /* stereo */
	float t_tone, t_inc, t;
#if SEND_COMP_CMD_TONE
	struct sof_ipc_ctrl_data *cmd;
	float a_tone, f_tone;
	int32_t frequency, amplitude;
#endif

	/* Handle command line */
	if (argc == 6) {
		fs1 = atof(argv[1]);
#if SEND_COMP_CMD_TONE
		f_tone = atof(argv[2]);
		a_tone = atof(argv[3]);
#endif
		t_tone = atof(argv[4]);
		fs2 = atof(argv[5]);
	} else {
		usage(argv[0]);
		return EXIT_SUCCESS;
	}

	/* Init components */
	sys_comp_init();
	sys_comp_tone_init();
	sys_comp_src_init();
	sys_comp_filewrite_init();

	/* Set SRC sink rate into IPC, src_new() will copy it */
	src_p0[0].sink_rate = fs2;
	deadline = pipeline[0].deadline;
	pipeline[0].frames_per_sched = (int) (0.9999 + fs2 * deadline / 1e6);

	/* Construct pipeline, pass parameters, start */
	if (test_bench_pipeline_setup(&reef, &pipeline[0], &spipe[0]) < 0)
		exit(EXIT_FAILURE);

	deadline = pipeline[0].deadline;
	if (test_bench_pipeline_params_start(reef.ipc, fs1, nch,
		deadline, SCHED_ID, PIPE_ID) < 0)
		exit(EXIT_FAILURE);

	/* Prepare for comp_cmd() */
	pcm_dev = ipc_get_comp(reef.ipc, TONE_ID);
	p = pcm_dev->cd->pipeline;
	cd = pcm_dev->cd;

#if SEND_COMP_CMD_TONE
	/* Frequency & amplitude command for steady tone*/
	frequency = (int32_t) (f_tone * 65536); /* Q18.14 */
	amplitude = (int32_t) (a_tone * 2147483647); /* Q1.31 */
	cmd = malloc(sizeof(struct sof_ipc_ctrl_data)
		+ sizeof(struct sof_abi_hdr)
		+TEST_BENCH_NCH * sizeof(struct sof_ipc_ctrl_value_comp));
	cmd->comp_id = TONE_ID;
	cmd->type = SOF_CTRL_TYPE_VALUE_COMP_SET;
	cmd->cmd = SOF_CTRL_CMD_ENUM;
	cmd->num_elems = TEST_BENCH_NCH;
	cmd->data->size = TEST_BENCH_NCH * sizeof(struct sof_ipc_ctrl_value_comp);
	cmd->data->abi = SOF_ABI_VERSION;
	cmd->data->magic = SOF_ABI_MAGIC;
	cmd->data->comp_abi = SOF_TONE_ABI_VERSION;
	//cdata = (struct sof_ipc_ctrl_value_comp *) cmd->data->data;

	/* Frequency */
	ret = tone_cmd_helper(cd, cmd, SOF_TONE_IDX_FREQUENCY, nch, frequency);
	if (ret < 0) {
		printf("Error: Tone frequency set failed!\n");
		exit(EXIT_FAILURE);
	}

	/* Amplitude */
	ret = tone_cmd_helper(cd, cmd, SOF_TONE_IDX_AMPLITUDE, nch, amplitude);
	if (ret < 0) {
		printf("Error: Tone amplitude set failed!\n");
		exit(EXIT_FAILURE);
	}
#endif
	/* Get filewrite state */
	pcm_dev = ipc_get_comp(reef.ipc, FW_ID);
	fwcd = comp_get_drvdata(pcm_dev->cd);

	/*
	 * Run copy until t_tone is reached
	 */
	pcm_dev = ipc_get_comp(reef.ipc, SCHED_ID);
	p = pcm_dev->cd->pipeline;
	cd = pcm_dev->cd;
	test_bench_disable_trace(); /* Reduce print output */
	tic = clock();

	t_inc = 1e-6 * deadline;
	t = t_inc;
	while (t < t_tone) {
		pipeline_schedule_copy(p, 0);
		t = t + t_inc;
	}

	/* Done! Next reset and free pipeline */
	toc = clock();
	test_bench_enable_trace();
	ret = pipeline_reset(p, cd);
	if (ret < 0) {
		printf("Failed pipeline_reset()\n");
		exit(EXIT_FAILURE);
	}
	pipeline_free(p);

	n_out = fwcd->fws.n;
	t_exec = (double) (toc - tic) / CLOCKS_PER_SEC;
	c_realtime = (double) n_out / nch / fs2 / t_exec;
	printf("wrote %d output samples,\n", n_out);
	printf("test execution time was %.2f us, %.2f x realtime.\n",
		1e3 * t_exec, c_realtime);

	return EXIT_SUCCESS;
}

void usage(char *executable)
{
	fprintf(stderr, "Usage: %s <tone rate> <frequency> <amplitude> <seconds> <out rate>\n",
		executable);
	fprintf(stderr, "E.g. \"%s 44100 997 0.5 1.0 48000\"\n", executable);
}

File: tune/src/tonesrc_test.m
function n_fail = tonesrc_test(fs_in_list, fs_out_list)

%%
% tonesrc_test - test tonesrc executable objective audio quality parameters
%
% src_test(fs_in, fs_out)
%
% fs_in  - vector of rates in
% fs_out - vector of rates out
%
% A default in-out matrix is tested if the paremeters are omitted.
%

%%
% Copyright (c) 2016, Intel Corporation
% All rights reserved.
%
% Redistribution and use in source and binary forms, with or without
% modification, are permitted provided that the following conditions are met:
%   * Redistributions of source code must retain the above copyright
%     notice, this list of conditions and the following disclaimer.
%   * Redistributions in binary form must reproduce the above copyright
%     notice, this list of conditions and the following disclaimer in the
%     documentation and/or other materials provided with the distribution.
%   * Neither the name of the Intel Corporation nor the
%     names of its contributors may be used to endorse or promote products
%     derived from this software without specific prior written permission.
%
% THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
% AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
% IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
% ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
% LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
% CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
% SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
% INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
% CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
% ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
% POSSIBILITY OF SUCH DAMAGE.
%
% Author: Seppo Ingalsuo <seppo.ingalsuo@linux.intel.com>
%

addpath('../m');
mkdir_check('plots');
mkdir_check('reports');

if nargin < 1
        fs_in_list = [8e3 16e3 24e3 32e3 44100 48e3 96e3 192e3];
end
if nargin < 2
        fs_out_list = fs_in_list;
end


%% Loop all modes to test
ch = 1;
nch = 2;
gtol_db = 0.1;
drlimit_db = 140;
n_fail = 0;
n_fsi = length(fs_in_list);
n_fso = length(fs_out_list);
dr = zeros(n_fsi, n_fso);
g = zeros(n_fsi, n_fso);
pf = zeros(n_fsi, n_fso);
for b = 1:n_fso
        for a = 1:n_fsi
                fs1 = fs_in_list(a);
                fs2 = fs_out_list(b);
                [pf(a,b), dr(a,b), g(a,b), missing] = dr_test(fs1, fs2, gtol_db, drlimit_db, ch, nch);
                n_fail = n_fail + pf(a,b);
                if missing
                        pf(a, b) = -1;
                end
        end
end

%% Print table with test summary: DR
fn = 'reports/dr_tonesrc.txt';
print_dr(fn, fs_in_list, fs_out_list, dr, pf);

%% Print table with test summary: Gain
fn = 'reports/g_tonesrc.txt';
print_g(fn, fs_in_list, fs_out_list, g, pf);

%% Print table with test summary: pass/fail
fn = 'reports/pf_tonesrc.txt';
print_pf(fn, fs_in_list, fs_out_list, pf);

fprintf('\n');
fprintf('Number of failed tests = %d\n', n_fail);

%% Delete src in/out text files
if n_fail == 0
        delete_check('tonesrc_out.txt');
end

end

function y = trim_sample(x, t_start, fs)
y = x(round(t_start*fs:end));
end

function [fail, dr_db, g_db, missing] = dr_test(fs1, fs2, gtol_db, ...
        mindr_db, ch, nch)

%% AES17 6.4.1 Dynamic range

%% Run SRC with sine input -60 dB relative to maximum input
fn_out = 'tonesrc_out.txt';
t = 3.0;
t_start = 1.0; % Ignore first 1.0s of SRC output notch filter to settle
f = 997;
a_db =-60; % As in AES17, though this low level hides quite a lot of aliasing
a = 10^(a_db/20);
bits = 32;
src_out = run_tonesrc(fs1, fs2, f, a, t, ch, nch);
if length(src_out) < 1
        fail = 0; dr_db = 0; g_db = 0; missing = 1;
        return;
else
        missing = 0;
end
src_out = src_out / 2^(bits-1);

%% Notch filter
y0n = stdnotch(src_out, f, fs2);

%% CCIR-RMS weighting filter
y0w = stdweight(y0n, fs2);

%% Trim sample by removing first 1s to let the notch to apply
i1 = round(t_start*fs2)+1;
y = src_out(i1:end);
y_n = y0w(i1:end);

%% Gain, SNR
level_in = a_db;
level_out = level_dbfs(y);
level_final = level_dbfs(y_n);
g_db = level_out-level_in;
dr_db = level_out-level_final-a_db;
fprintf('Gain = %4.1f dB, DR = %5.1f dB\n', g_db, dr_db);
if 0
        plot(y);
        hold on
        plot(y_n,'g');
        hold off
end

%% Check pass/fail
fail = 0;
if abs(g_db) > gtol_db
        fprintf('Failed gain %f dB (max %f dB)\n', g_db, gtol_db);
        fail = 1;
end
if dr_db < mindr_db
        fprintf('Failed DR %f dB (min %f dB)\n', dr_db, mindr_db);
        fail = 1;
end

end


function [out, in] = run_tonesrc(fs1, fs2, f, a, t, ch, nch)
ex = './tonesrc_test';
fn_in = 'tonesrc_in.txt';
fn_out = 'tonesrc_out.txt';
src_cmd = sprintf('%s %d %f %f %f %d', ex, fs1, f, a, t, fs2);
fprintf('%s\n', src_cmd);
if exist(ex) == 2
else
        error('Can''t find compiled executable %s', ex);
end
if (ch < 1)
        ch = 1+round(rand(1,1)*(nch-1));
end
delete_check(fn_out);
system(src_cmd);
out = [];
if exist(fn_out)
	dir_fn_out = dir(fn_out);
	if dir_fn_out.bytes > 0
		out = load(fn_out);
		out = out(ch:nch:end);
	end
end

if exist(fn_in)
        in = load(fn_in);
        in = in(ch:nch:end);
else
        in = [];
end

end



function  print_dr(fn, fs_in_list, fs_out_list, dr, pf);

n_fsi = length(fs_in_list);
n_fso = length(fs_out_list);
fh = fopen(fn,'w');
fprintf(fh,'\n');
fprintf(fh,'SRC test result: DR dB(CCIR-RMS)\n');
fprintf(fh,'%8s, ', 'out \ in');
for a = 1:n_fsi-1
        fprintf(fh,'%8.1f, ', fs_in_list(a)/1e3);
end
fprintf(fh,'%8.1f', fs_in_list(n_fsi)/1e3);
fprintf(fh,'\n');
for b = 1:n_fso
        fprintf(fh,'%8.1f, ', fs_out_list(b)/1e3);
        for a = 1:n_fsi
                if pf(a,b,1) < 0
                        cstr = 'x';
                else
                        cstr = sprintf('%5.1f', dr(a,b));
                end
                if a < n_fsi
                        fprintf(fh,'%8s, ', cstr);
                else
                        fprintf(fh,'%8s', cstr);
                end
        end
        fprintf(fh,'\n');
end
fclose(fh);
type(fn);
end

function  print_g(fn, fs_in_list, fs_out_list, g, pf);

n_fsi = length(fs_in_list);
n_fso = length(fs_out_list);
fh = fopen(fn,'w');
fprintf(fh,'\n');
fprintf(fh,'SRC test result: Gain dB\n');
fprintf(fh,'%8s, ', 'out \ in');
for a = 1:n_fsi-1
        fprintf(fh,'%8.1f, ', fs_in_list(a)/1e3);
end
fprintf(fh,'%8.1f', fs_in_list(n_fsi)/1e3);
fprintf(fh,'\n');
for b = 1:n_fso
        fprintf(fh,'%8.1f, ', fs_out_list(b)/1e3);
        for a = 1:n_fsi
                if pf(a,b,1) < 0
                        cstr = 'x';
                else
                        cstr = sprintf('%8.5f', g(a,b));
                end
                if a < n_fsi
                        fprintf(fh,'%8s, ', cstr);
                else
                        fprintf(fh,'%8s', cstr);
                end
        end
        fprintf(fh,'\n');
end
fclose(fh);
type(fn);
end


function print_pf(fn, fs_in_list, fs_out_list, pf)

n_fsi = length(fs_in_list);
n_fso = length(fs_out_list);
fh = fopen(fn,'w');
fprintf(fh,'\n');
fprintf(fh,'SRC test result: Fails\n');
fprintf(fh,'%8s, ', 'out \ in');
for a = 1:n_fsi-1
        fprintf(fh,'%8.1f, ', fs_in_list(a)/1e3);
end
fprintf(fh,'%8.1f', fs_in_list(n_fsi)/1e3);
fprintf(fh,'\n');
for b = 1:n_fso
        fprintf(fh,'%8.1f, ', fs_out_list(b)/1e3);
        for a = 1:n_fsi
                if pf(a,b,1) < 0
                        cstr = 'x';
                else
                        cstr = sprintf('%d', pf(a,b,:));
                end
                if a < n_fsi
                        fprintf(fh,'%8s, ', cstr);
                else
                        fprintf(fh,'%8s', cstr);
                end
        end
        fprintf(fh,'\n');
end
fclose(fh);
type(fn);
end
